File: MtClient/MtRpcClient.cs
using System.Net.WebSockets;
using System.Text;

namespace MtClient
{
    public interface IRpcLogger
    {
        public void Debug(string message);
        public void Info(string message);
        public void Warn(string message);
        public void Error(string message);
    }

    public class MtRpcClient
    {
        public MtRpcClient(string host, int port, IRpcLogger? logger = null)
        {
            host_ = host;
            port_ = port;
            logger_ = logger ?? new StubLogger();

            receiveThread_ = new Thread(new ThreadStart(DoReceive));
            sendThread_ = new Thread(new ThreadStart(DoWrite));
        }

        public async Task Connect()
        {
            logger_.Debug($"MtRpcClient.Connect: started to {host_}:{port_}");

            try
            {
                await ws_.ConnectAsync(new Uri($"ws://{host_}:{port_}/ws"), CancellationToken.None);
            }
            catch (Exception ex)
            {
                logger_.Error($"MtRpcClient.Connect failed: {ex.Message}");
                throw new Exception($"Failed connection to {host_}:{port_}");
            }

            receiveThread_.Start();
            sendThread_.Start();

            logger_.Debug($"MtRpcClient.Connect: success.");
        }

        public async void Disconnect()
        {
            logger_.Debug($"MtRpcClient.Disconnect: {host_}:{port_}");

            try
            {
                if (ws_.State == WebSocketState.Open)
                    await ws_.CloseAsync(WebSocketCloseStatus.NormalClosure, null, CancellationToken.None);
                ws_.Dispose();
            }
            catch (Exception ex)
            {
                logger_.Warn($"MtRpcClient.Disconnect: {ex.Message}");
            }

            sendWaiter_.Set();
            sendThread_.Join();
            receiveThread_.Join();

            logger_.Debug($"MtRpcClient.Disconnect: success");
        }

        public string? SendCommand(int expertHandle, int commandType, string payload)
        {
            CommandTask commandTask = new();
            int commandId;
            lock (tasks_)
            {
                commandId = nextCommandId++;
                tasks_[commandId] = commandTask;
            }

            MtCommand command = new(expertHandle, commandType, commandId, payload);
            Send(command);

            var response = commandTask.WaitResponse(10000); // 10 sec
            lock (tasks_)
            {
                tasks_.Remove(commandId);
            }

            return response;
        }

        public void NotifyClientReady()
        {
            MtNotification notification = new(MtNotificationType.ClientReady);
            Send(notification);
        }

        private void Send(MtMessage message)
        {
            lock (pendingMessages_)
            {
                pendingMessages_.Enqueue(message);
            }
            sendWaiter_.Set();
        }

        private async void DoWrite()
        {
            while(ws_.State == WebSocketState.Open)
            {
                MtMessage? message = null;
                lock(pendingMessages_)
                {
                    if (pendingMessages_.Count > 0)
                        message = pendingMessages_.Dequeue();
                }

                if (message == null)
                {
                    sendWaiter_.WaitOne();
                    continue;
                }                   

                try
                {
                    string msgStr = message.Serialize();
                    logger_.Debug($"MtRpcClient.DoWrite: sending message: {msgStr}");
                    byte[] bytes = Encoding.ASCII.GetBytes(msgStr);
                    await ws_.SendAsync(bytes, WebSocketMessageType.Text, true, CancellationToken.None);
                }
                catch (Exception e)
                {
                    logger_.Error($"MtRpcClient.DoWrite: {e.Message}");
                }
            }
        }

        private async void DoReceive()
        {
            try
            {
                byte[] recvBuffer = new byte[64 * 1024];
                while (ws_.State == WebSocketState.Open)
                {
                    var result = await ws_.ReceiveAsync(new ArraySegment<byte>(recvBuffer), CancellationToken.None);
                    if (result.MessageType == WebSocketMessageType.Close)
                    {
                        logger_.Info($"MtRpcClient.DoReceive: close signal {result.CloseStatusDescription}");
                        Disconnected?.Invoke(this, EventArgs.Empty);
                        break;
                    }
                    else
                    {
                        var msg = Encoding.ASCII.GetString(recvBuffer, 0, result.Count);
                        OnReceive(msg);
                    }
                }
            }
            catch (Exception ex)
            {
                logger_.Warn($"MtRpcClient.DoReceive: Exception in receive - {ex.Message}");
                ConnectionFailed?.Invoke(this, EventArgs.Empty);
            }
        }

        private void OnReceive(string msg)
        {
            logger_.Debug($"MtRpcClient.OnReceive: {msg}");

            if (string.IsNullOrEmpty(msg))
            {
                logger_.Warn("MtRpcClient.OnReceive: Invalid message (null or empty)");
                return;
            }

            var pieces = msg.Split(";", 2);

            if (pieces.Length != 2
                || string.IsNullOrEmpty(pieces[0])
                || string.IsNullOrEmpty(pieces[1]))
            {
                logger_.Warn("MtRpcClient.OnReceive: Invalid message format.");
                return;
            }                

            MessageType msgType;
            try
            {
                var msgTypeValue = int.Parse(pieces[0]);
                msgType = (MessageType)Enum.ToObject(typeof(MessageType), msgTypeValue);

            }
            catch (Exception e)
            {
                logger_.Error($"MtRpcClient.OnReceive: Parse MessageType failed. {e.Message}");
                return;
            }

            var message = MtMessageParser.Parse(msgType, (pieces[1]));
            if (message == null)
            {
                logger_.Error("MtRpcClient.OnReceive: Failed parse message payload");
                return;
            }

            switch (message.MsgType)
            {
                case MessageType.Event:
                    if (message is MtEvent e)
                        MtEventReceived?.Invoke(this, new(e.ExpertHandle, e.EventType, e.Payload));
                    break;
                case MessageType.Response:
                    if (message is MtResponse response)
                        ProcessResponse(response);
                    break;
                case MessageType.ExpertList:
                    if (message is MtExpertListMsg expertListMsg)
                        ExpertList?.Invoke(this, new(expertListMsg.Experts));
                    break;
                case MessageType.ExpertAdded:
                    if (message is MtExpertAddedMsg expertAddedMsg)
                        ExpertAdded?.Invoke(this, new(expertAddedMsg.ExpertHandle));
                    break;
                case MessageType.ExpertRemoved:
                    if (message is MtExpertRemovedMsg expertRemovedMsg)
                        ExpertRemoved?.Invoke(this, new(expertRemovedMsg.ExpertHandle));
                    break;
            }
        }

        private void ProcessResponse(MtResponse msg)
        {
            var handle = msg.ExpertHandle;
            var commandId = msg.CommandId;
            var payload = msg.Payload;

            logger_.Debug($"MtRpcClient.ProcessResponse: {handle}, {commandId}, [{payload}]");

            lock (tasks_)
            {
                if (tasks_.TryGetValue(commandId, out CommandTask? value))
                    value.SetResponse(payload);
            }
        }

        public event EventHandler<EventArgs>? ConnectionFailed;
        public event EventHandler<EventArgs>? Disconnected;
        public event EventHandler<MtExpertListEventArgs>? ExpertList;
        public event EventHandler<MtExpertEventArgs>? ExpertAdded;
        public event EventHandler<MtExpertEventArgs>? ExpertRemoved;
        public event EventHandler<MtEventArgs>? MtEventReceived;

        private readonly ClientWebSocket ws_ = new();
        private readonly string host_;
        private readonly int port_;
        private readonly byte[] buf_ = new byte[10000];
        private readonly Queue<MtMessage> pendingMessages_ = [];

        private readonly Thread receiveThread_;
        private readonly Thread sendThread_;
        private readonly EventWaitHandle sendWaiter_ = new AutoResetEvent(false);

        private int nextCommandId = 0;
        private readonly Dictionary<int, CommandTask> tasks_ = [];

        private readonly IRpcLogger logger_;
    }

    internal class CommandTask
    {
        private readonly EventWaitHandle responseWaiter_ = new AutoResetEvent(false);
        private string? response_;
        private readonly object locker_ = new();

        public string? WaitResponse(int time)
        {
            responseWaiter_.WaitOne(time);
            lock (locker_)
            {
                return response_;
            }
        }

        public void SetResponse(string result)
        {
            lock (locker_)
            {
                response_ = result;
            }
            responseWaiter_.Set();
        }
    }

    public class MtEventArgs(int expertHandle, int eventType, string payload) : EventArgs
    {
        public int ExpertHandle { get; } = expertHandle;
        public int EventType { get; } = eventType;
        public string Payload { get; } = payload;
    }

    public class MtExpertListEventArgs(HashSet<int> experts) : EventArgs
    {
        public HashSet<int> Experts { get; } = experts;
    }

    public class MtExpertEventArgs(int expert) : EventArgs
    {
        public int Expert { get;  }= expert;
    }

    internal class StubLogger : IRpcLogger
    {
        public void Debug(string message)
        {
        }

        public void Error(string message)
        {
        }

        public void Info(string message)
        {
        }

        public void Warn(string message)
        {
        }
    }
}

File: PyMtApi5/mtrpcclient.py
import logging
from enum import IntEnum
from threading import Condition, Lock, Thread

import websockets
from websockets.sync.client import connect as ws_connect


class MtNotification(IntEnum):
    ClientReady = 0

class MtMessageType(IntEnum):
    Command = 0
    Response = 1
    Event = 2
    ExpertList = 3
    ExpertAdded = 4
    ExpertRemoved = 5
    Notification = 6

class CommandTask:
    def __init__(self):
        self.locker = Lock()
        self.waiter = Condition()
        self.response = None

    def wait_response(self, time):
        with self.waiter:
            self.waiter.wait(time)
        with self.locker:
            return self.response

    def set_response(self, response):
        with self.locker:
            self.response = response
        with self.waiter:
            self.waiter.notify()

class MtRpcClient:
    def __init__(self, callback=None):
        self.__logger = logging.getLogger(__name__)
        self.__callback = callback
        self.__notification_tasks = dict()
        self.__tasks = dict()
        self.__next_command_id = 0
        self.__lock = Lock()

    def connect(self, url):
        self.__logger.debug(f"connecting to {url}")
        self.__ws = ws_connect(url);
        self.__receive_thread = Thread(target = self.__receive_messages_thread)
        self.__receive_thread.start()

    def disconnect(self):
        self.__ws.close()
        self.__receive_thread.join()
        self.__logger.debug("disconnected")

    def request_expert_list(self):
        task = CommandTask()
        with self.__lock:
            self.__notification_tasks[MtNotification.ClientReady] = task
        self.__ws.send(self.__create_notification(MtNotification.ClientReady))
        response = task.wait_response(10)
        with self.__lock:
            self.__notification_tasks.pop(MtNotification.ClientReady)
        return response

    def send_command(self, expert_handle, command_type, payload = None):
        command_id = self.__next_command_id
        self.__next_command_id += 1
        task = CommandTask()
        with self.__lock:
            self.__tasks[command_id] = task
        self.__ws.send(self.__create_mt_command(expert_handle, command_id, command_type, payload))
        response = task.wait_response(10)
        with self.__lock:
            self.__tasks.pop(command_id)
        return response

    # Private methods

    def __process_message(self, message):
        self.__logger.debug(f"process_message: {message}")
        pieces = message.split(';', 1)
        if len(pieces) != 2 or not pieces[0] or not pieces[1]:
            self.__logger.warning("process_message: Invalid message format");
            return
        message_type = MtMessageType(int(pieces[0]))
        if message_type == MtMessageType.ExpertList:
            self.__process_expert_list(pieces[1])
        elif message_type == MtMessageType.Event:
            self.__process_event(pieces[1])
        elif message_type == MtMessageType.Response:
            self.__process_response(pieces[1])
        elif message_type == MtMessageType.ExpertAdded:
            self.__process_expert_added(pieces[1])
        elif message_type == MtMessageType.ExpertRemoved:
            self.__process_expert_removed(pieces[1])
        else:
            self.__logger.warning(f"received unknown message type: {message_type}")

    def __process_expert_list(self, payload):
        pieces = payload.split(',')
        experts = list()
        for p in pieces:
            experts.append(int(p))
        with self.__lock:
            task = self.__notification_tasks.get(MtNotification.ClientReady)
            if task is not None:
                task.set_response(experts)

    def __process_event(self, payload):
        pieces = payload.split(';', 2)
        if len(pieces) != 3 or not pieces[0] or not pieces[1] or not pieces[2]:
            self.__logger.warning("process_event: Invalid message format");
            return
        if self.__callback is not None:
            self.__callback.mt_rpc_on_event(int(pieces[0]), int(pieces[1]), pieces[2])

    def __process_response(self, payload):
        pieces = payload.split(';', 2)
        if len(pieces) != 3 or not pieces[0] or not pieces[1] or not pieces[2]:
            self.__logger.warning("process_response: Invalid message format");
            return
        command_id = int(pieces[1])
        with self.__lock:
            task = self.__tasks.get(command_id)
            if task is not None:
                task.set_response(pieces[2])

    def __process_expert_added(self, payload):
        if self.__callback is not None:
            self.__callback.mt_rpc_on_expert_added(int(payload))

    def __process_expert_removed(self, payload):
        if self.__callback is not None:
            self.__callback.mt_rpc_on_expert_removed(int(payload))

    def __receive_messages_thread(self):
        self.__logger.debug("started receive_messages thread")
        while True:
            try:
                message = self.__ws.recv()
                self.__process_message(message)
            except websockets.exceptions.ConnectionClosed:
                self.__logger.info("Connection closed")
                if self.__callback is not None:
                    self.__callback.mt_rcp_on_disconnect()
                break
            except Exception as e:
                self.__logger.error(e)
                if self.__callback is not None:
                    self.__callback.mt_rpc_on_connection_failed(str(e))
                break
        self.__logger.debug("function receive_messages finished")

    def __create_notification(self, notification_type):
        return f"{int(MtMessageType.Notification)};{notification_type}"

    def __create_mt_command(self, expert_handle, command_id, command_type, payload):
        if (payload is None):
            return f"{MtMessageType.Command};{expert_handle};{command_id};{command_type}";
        return f"{MtMessageType.Command};{expert_handle};{command_id};{command_type};{payload}";
